inst/cpp/sim.cpp

```diff
@@ -63,23 +63,116 @@ void pk_code (int i, std::vector<double> times, std::vector<double> doses, doubl
+    lagtime = NumericVector(n_comp);
+    for(int i = 0; i < current_size; i++) {
+      lagtime[i] = tlag[i];
+    }
+    for(int i = current_size; i < n_comp; i++) {
+      lagtime[i] = 0.0;
+    }
+  } else {
+    lagtime = tlag;
+  }
+
+  // Apply lagtime to dose events (evid == 1)  
+  for(int i = 0; i < times.size(); i++) {
+    if(evid[i] == 1) {
+      times[i] += lagtime[cmt[i]-1];
+    }
+  }
+
+  // Create sorted index according to "t"
+  std::vector<size_t> indices(times.size());
+  std::iota(indices.begin(), indices.end(), 0); // Fill with 0, 1, 2, ...
+  std::sort(indices.begin(), indices.end(), [&times](size_t i1, size_t i2) {
+    return times[i1] < times[i2];
+  });
+  
+  // Reorder all elements in `t` in new_design
+  std::vector<double> sorted_times(times.size());
+  for (size_t i = 0; i < indices.size(); i++) {
+    sorted_times[i] = times[indices[i]];
+  }
+  new_design["t"] = sorted_times;
+  
+  // Sort all other vectors in the design object
+  for (const char* key : {"dose", "type", "dum", "dose_cmt", "t_inf", "evid", "bioav", "rate", "obs_type"}) {
+    if (new_design.containsElementNamed(key)) {
+      SEXP vec = new_design[key];
+      if (TYPEOF(vec) == REALSXP) {
+        std::vector<double> old_vec = as<std::vector<double> >(vec);
+        std::vector<double> new_vec(old_vec.size());
+        for (size_t i = 0; i < indices.size(); i++) {
+          new_vec[i] = old_vec[indices[i]];
+        }
+        new_design[key] = new_vec;
+      } else if (TYPEOF(vec) == INTSXP) {
+        std::vector<int> old_vec = as<std::vector<int> >(vec);
+        std::vector<int> new_vec(old_vec.size());
+        for (size_t i = 0; i < indices.size(); i++) {
+          new_vec[i] = old_vec[indices[i]];
+        }
+        new_design[key] = new_vec;
+      }
+    }
+  }
+  
+  return new_design;
+}
+
 // [[Rcpp::export]]
-List sim_wrapper_cpp (NumericVector A, List design, List par, NumericVector iov_bins, double step_size) {
+List sim_wrapper_cpp (NumericVector A, List design, List par, NumericVector iov_bins, SEXP lagtime, double step_size) {
   std::vector<double> t;
   std::vector<state_type> y;
   // insert observation variable definition
   double t_start, t_end;
   std::vector<double> times, doses, dummy, rates;
   std::vector<int> dose_cmt, dose_type, evid, obs_type, y_type;
   // insert variable definitions
-  times = as<std::vector<double> >(design["t"]);
-  doses = as<std::vector<double> >(design["dose"]);
-  evid = as<std::vector<int> >(design["evid"]);
-  dummy = as<std::vector<double> >(design["dum"]);
-  rates = as<std::vector<double> >(design["rate"]);
-  dose_cmt = as<std::vector<int> >(design["dose_cmt"]);
-  dose_type = as<std::vector<int> >(design["type"]);
-  obs_type = as<std::vector<int> >(design["obs_type"]);
+
+  // Handle lagtime parameter - can be numeric or character
+  NumericVector lagtime_numeric = lagtime_to_numeric(lagtime, par);
+  List events = apply_lagtime(design, lagtime_numeric, n_comp);
```

Comment:
first convert the vector to numeric (e.g. from possible parameter "TLAG"), then apply the lagtime vector to the event table.

Reply:
where do we pull n_comp from?

Reply:
it's a global variable that's added dynamically before the file gets compiled. Not ideal that it is global, but we have a ticket to address ([RXR-2712](https://insight-rx.atlassian.net/browse/RXR-2712).)